Automattic/wp-calypso
# `/read/subscriptions`: Implement user interactions, tracking, and loading state for the new `RecommendedSites` component

![Screen Capture on 2023-06-11 at 10-40-58](https://github.com/Automattic/wp-calypso/assets/2019970/4636edd5-44c9-4416-8746-4d0f11a59063)



Closes https://github.com/Automattic/wp-calypso/issues/77890 
Closes https://github.com/Automattic/wp-calypso/issues/77891 
Closes https://github.com/Automattic/wp-calypso/issues/77893 

## Proposed Changes

Apologies for the confusion earlier. Here's a more comprehensive simplification, maintaining all the specific tracking events:

1. Developed `RecommendedSitesPlaceholder` and `RecommendedSitePlaceholder` to show loading status of recommended sites. This follows the strategy of the older `client/blocks/recommended-sites` component.

2. Retained the same statistics and analytics triggers as the previous RecommendedSites component:

    - When the RecommendedSite is displayed or changes position in the UI:
      - Triggers a 'rail car rendered' event for `recommended_site`, including `ui_algo` and `ui_position` properties.
      - The position might change if the adjacent recommended site is either subscribed to or dismissed, prompting the current site to move up.

    - When a recommended site is dismissed via 'X' button:
      - Records a `calypso_reader_recommended_site_dismissed` event with `railcar` object and `ui_positions` property.
      - Increases `reader_actions` stat with `calypso_reader_recommended_site_dismissed` as the action property.

    - When a site avatar is clicked:
      - Records a `calypso_reader_avatar_clicked` event with properties such as `blog_id`, `feed_id`, and `source` as `reader-following-manage-recommendation`.
      - Triggers a train-tracks event with the same name and properties.

    - When a site title is clicked:
      - Records a `calypso_reader_feed_link_clicked` event with properties such as `blog_id`, `feed_id`, and `source` as `reader-following-manage-recommendation`.
      - Triggers a train-tracks event with the same name and properties.

    - When a site URL is clicked:
      - Records a `calypso_reader_site_url_clicked` event with properties such as `blog_id`, `feed_id`, and `source` as `reader-following-manage-recommendation`.
      - Triggers a train-tracks event with the same name and properties.

    - When "Subscribe" is clicked:
      - Increases `reader_follows` stat with `reader-following-manage-recommendation` as a parameter.
      - Increases `reader_actions` stat with `followed_blog` as a parameter.
      - Records a Google Analytics event with `Reader` category, `Clicked Follow Blog` action, and `reader-following-manage-recommendation` label.
      - Records a `calypso_reader_site_followed` event with URL and source properties.
      - Records a `calypso_traintracks_interact` event, setting event name to 'site_followed' and passing the appropriate railcar object.

3. Introduced `reader.follows.followFeedLoading`, a Redux state variable representing `feedUrl`s with pending follow requests. This enables the UI to indicate the status of a specific follow request. Once the request is successful or fails, the `feedUrl` is removed from the array.

    - **Potential improvement idea**: to utilize a `keyedReducer` where `feedUrl` is the key. This could change the state to be boolean with a default `false`.
    - Added `isReaderFollowFeedLoading` selector for the new state variable.
    
4. Introduced two new actions, `READER_RECOMMENDED_SITE_DISMISSED` and `READER_RECOMMENDED_SITE_FOLLOWED`. These are dispatched only after a site has been successfully dismissed or followed, and if the `reader/subscription-management` flag is enabled. These actions help in removing the affected site’s entry from the `state.reader.recommended-sites.items` list that's used for rendering the recommended sites to the user.


____

### Updating the site subscriptions list after Subscribe on a recommended site
**⚠️ This PR doesn't include the implementation of updating the subscriptions list after successfully subscribing to a recommended site.**

   Though it may appear to be a straightforward task, it ideally requires the creation of a dedicated "SiteSubscriptionsManagerProvider" for Reader.
    
   As of now, the site subscriptions list is fetched using react-query, while the whole reader architecture is built on Redux and wpcom/data-layer middleware. It would make the most sense to create a separate provider for the reader, one that provides data and mutation handlers from Redux and handles dispatch actions.
    
   However, the provider for the external subscriptions management portal should continue to rely on react-query and the context itself for data storage.
    
   To summarize, this PR doesn't involve updating the subscriptions list post successful site subscription nor does it establish the new "SiteSubscriptionsManagerProvider". These are expected to be addressed separately.
____

## Testing Instructions



* Go to http://calypso.localhost:3000/read/subscriptions, and play around with the recommended cards.
  * Ensure all the required analytics and stats calls are made. See the above description for the full list.
* Go to http://calypso.localhost:3000/following/manage, and test for any regressions.

## Pre-merge Checklist
